File: go/io-benchmark-heavy.go
package main

import (
	"bytes"
	"compress/gzip"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"os"
	"strings"
	"sync"
	"time"
)

// FileData represents the structure of our test files
type FileDataHeavy struct {
	Hash      string `json:"hash"`
	Data      string `json:"data"`
	Timestamp int64  `json:"timestamp"`
}

// Generate large file content (several MB)
// Go handles this efficiently with string builder and native string operations
func generateLargeContentGo(sizeMB int, seed string) string {
	var builder strings.Builder
	chunkSize := 1024 // 1KB chunks
	totalChunks := sizeMB * 1024
	
	builder.Grow(totalChunks * chunkSize) // Pre-allocate memory (Go advantage)
	
	for i := 0; i < totalChunks; i++ {
		// Create deterministic content
		chunk := fmt.Sprintf("Data chunk %d with seed %s: %s\n", i, seed, strings.Repeat("x", 900))
		builder.WriteString(chunk)
	}
	
	return builder.String()
}

// CPU-intensive: Compute SHA256 hash
// Go can run multiple goroutines in parallel across CPU cores
func computeHashGo(content string) string {
	hash := sha256.Sum256([]byte(content))
	return hex.EncodeToString(hash[:])
}

// CPU-intensive: Compress data using gzip
// Go can parallelize multiple compression tasks across goroutines
func compressDataGo(content string) ([]byte, error) {
	var buf bytes.Buffer
	writer := gzip.NewWriter(&buf)
	
	_, err := writer.Write([]byte(content))
	if err != nil {
		return nil, err
	}
	
	err = writer.Close()
	if err != nil {
		return nil, err
	}
	
	return buf.Bytes(), nil
}

// Decompress gzip data
func decompressDataGo(compressed []byte) (string, error) {
	reader, err := gzip.NewReader(bytes.NewReader(compressed))
	if err != nil {
		return "", err
	}
	defer reader.Close()
	
	decompressed, err := io.ReadAll(reader)
	if err != nil {
		return "", err
	}
	
	return string(decompressed), nil
}

func createTestFilesHeavy(count int, taskID string, sizeMB int) error {
	var wg sync.WaitGroup
	errChan := make(chan error, count)
	
	for i := 0; i < count; i++ {
		wg.Add(1)
		// Go can run all these goroutines in parallel across CPU cores
		go func(i int) {
			defer wg.Done()
			
			// Generate large content (parallel across goroutines)
			content := generateLargeContentGo(sizeMB, fmt.Sprintf("%s_%d", taskID, i))
			
			// Compute hash (parallel across goroutines)
			hash := computeHashGo(content)
			
			// Compress data (parallel across goroutines)
			compressed, err := compressDataGo(content)
			if err != nil {
				errChan <- err
				return
			}
			
			// Create file data structure
			fileData := FileDataHeavy{
				Hash:      hash,
				Data:      hex.EncodeToString(compressed),
				Timestamp: time.Now().UnixMilli(),
			}
			
			// Write to file
			jsonData, err := json.Marshal(fileData)
			if err != nil {
				errChan <- err
				return
			}
			
			filename := fmt.Sprintf("temp_%s_%d.dat", taskID, i)
			if err := os.WriteFile(filename, jsonData, 0644); err != nil {
				errChan <- err
			}
		}(i)
	}
	
	wg.Wait()
	close(errChan)
	
	if len(errChan) > 0 {
		return <-errChan
	}
	return nil
}

func readTestFilesHeavy(count int, taskID string) ([]string, error) {
	var wg sync.WaitGroup
	results := make([]string, count)
	errChan := make(chan error, count)
	
	for i := 0; i < count; i++ {
		wg.Add(1)
		// Go runs these in parallel across CPU cores
		go func(i int) {
			defer wg.Done()
			filename := fmt.Sprintf("temp_%s_%d.dat", taskID, i)
			
			// Retry logic
			var jsonData []byte
			var err error
			for retry := 0; retry < 10; retry++ {
				jsonData, err = os.ReadFile(filename)
				if err == nil {
					break
				}
				if retry < 9 {
					time.Sleep(20 * time.Millisecond)
				}
			}
			
			if err != nil {
				errChan <- err
				return
			}
			
			// Parse JSON
			var fileData FileDataHeavy
			if err := json.Unmarshal(jsonData, &fileData); err != nil {
				errChan <- err
				return
			}
			
			// Decode hex data
			compressed, err := hex.DecodeString(fileData.Data)
			if err != nil {
				errChan <- err
				return
			}
			
			// Decompress (parallel across goroutines)
			content, err := decompressDataGo(compressed)
			if err != nil {
				errChan <- err
				return
			}
			
			// Verify hash (parallel across goroutines)
			hash := computeHashGo(content)
			if hash != fileData.Hash {
				errChan <- fmt.Errorf("hash mismatch")
				return
			}
			
			results[i] = content
		}(i)
	}
	
	wg.Wait()
	close(errChan)
	
	if len(errChan) > 0 {
		return nil, <-errChan
	}
	return results, nil
}

func cleanupTestFilesHeavy(count int, taskID string) {
	var wg sync.WaitGroup
	for i := 0; i < count; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			filename := fmt.Sprintf("temp_%s_%d.dat", taskID, i)
			os.Remove(filename) // Ignore errors
		}(i)
	}
	wg.Wait()
}

// TaskResultHeavy holds timing information for a task
type TaskResultHeavy struct {
	FilesProcessed int
	TaskTime       time.Duration
}

func ioIntensiveTaskHeavy(taskID string) (TaskResultHeavy, error) {
	// Reduced file count but increased file size for CPU-intensive operations
	// Go excels here with parallel goroutines across multiple CPU cores
	fileCount := 10  // Fewer files but larger and more CPU-intensive
	fileSizeMB := 2  // 2MB per file (total 20MB per task)
	networkCalls := 5
	
	taskStart := time.Now()
	
	// Create large files with compression and hashing
	// Go runs these in parallel across available CPU cores
	if err := createTestFilesHeavy(fileCount, taskID, fileSizeMB); err != nil {
		return TaskResultHeavy{}, err
	}
	
	// Short wait to ensure filesystem writes complete
	time.Sleep(5 * time.Millisecond)
	
	// Simulate network calls with delays
	var wg sync.WaitGroup
	for i := 0; i < networkCalls; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			delay := 10 + rand.Intn(20)
			simulateNetworkDelay(delay)
		}()
	}
	
	// Read, decompress, and verify files (all in parallel)
	fileContents, err := readTestFilesHeavy(fileCount, taskID)
	wg.Wait() // Wait for network calls to complete
	
	// Cleanup
	cleanupTestFilesHeavy(fileCount, taskID)
	
	taskEnd := time.Now()
	
	if err != nil {
		return TaskResultHeavy{}, err
	}
	
	return TaskResultHeavy{
		FilesProcessed: len(fileContents),
		TaskTime:       taskEnd.Sub(taskStart),
	}, nil
}

func RunIOBenchmarkHeavy(taskCount int) {
	fmt.Println("Go Heavy I/O benchmark start")
	fmt.Println("WHY GO IS FASTER:")
	fmt.Println("1. True parallelism with goroutines across multiple CPU cores")
	fmt.Println("2. Efficient memory management with pre-allocation and minimal GC pressure")
	fmt.Println("3. Native compiled code for CPU-intensive operations (hash, compression)")
	fmt.Println("4. Concurrent I/O operations don't block each other")
	fmt.Println("---")

	start := time.Now()
	
	results := make([]TaskResultHeavy, taskCount)
	var mu sync.Mutex
	var wg sync.WaitGroup
	
	for i := 0; i < taskCount; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			taskID := fmt.Sprintf("task%d", i)
			result, err := ioIntensiveTaskHeavy(taskID)
			if err != nil {
				fmt.Printf("Task %d failed: %v\n", i+1, err)
			} else {
				mu.Lock()
				results[i] = result
				mu.Unlock()
				fmt.Printf("Task %d done in %v\n", i+1, result.TaskTime)
			}
		}(i)
	}

	wg.Wait()
	elapsed := time.Since(start)
	
	// Calculate statistics
	var totalTaskTime time.Duration
	var totalFiles int
	for _, r := range results {
		totalTaskTime += r.TaskTime
		totalFiles += r.FilesProcessed
	}
	avgTaskTime := totalTaskTime / time.Duration(len(results))
	
	fmt.Println("---")
	fmt.Printf("All %d tasks done in %v\n", taskCount, elapsed)
	fmt.Printf("Average task time: %v\n", avgTaskTime)
	fmt.Printf("Total files processed: %d\n", totalFiles)
}

File: rust/src/io_benchmark_heavy.rs
// Heavy I/O benchmark with CPU-intensive operations
// Demonstrates Rust's advantages: true parallelism, zero-cost abstractions, and efficient memory management
use std::time::{Duration, Instant};
use tokio::time::sleep;
use sha2::{Sha256, Digest};
use flate2::write::GzEncoder;
use flate2::read::GzDecoder;
use flate2::Compression;
use std::io::{Write, Read};
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
struct FileData {
    hash: String,
    data: String,
    timestamp: u128,
}

// Generate large file content (several MB)
// Rust handles this efficiently with String::with_capacity and zero-copy operations
fn generate_large_content(size_mb: usize, seed: &str) -> String {
    let chunk_size = 1024; // 1KB chunks
    let total_chunks = size_mb * 1024;
    
    // Pre-allocate memory (Rust advantage: precise memory control)
    let mut content = String::with_capacity(total_chunks * chunk_size);
    
    for i in 0..total_chunks {
        // Create deterministic content
        let chunk = format!("Data chunk {} with seed {}: {}\n", i, seed, "x".repeat(900));
        content.push_str(&chunk);
    }
    
    content
}

// CPU-intensive: Compute SHA256 hash
// Rust can parallelize across threads without runtime overhead
fn compute_hash(content: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(content.as_bytes());
    let result = hasher.finalize();
    hex::encode(result)
}

// CPU-intensive: Compress data using gzip
// Rust's zero-cost abstractions mean no performance penalty
fn compress_data(content: &str) -> Result<Vec<u8>, std::io::Error> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(content.as_bytes())?;
    encoder.finish()
}

// Decompress gzip data
fn decompress_data(compressed: &[u8]) -> Result<String, std::io::Error> {
    let mut decoder = GzDecoder::new(compressed);
    let mut decompressed = String::new();
    decoder.read_to_string(&mut decompressed)?;
    Ok(decompressed)
}

async fn create_test_files(
    count: usize, 
    task_id: &str, 
    size_mb: usize
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut handles = Vec::new();
    let task_id = task_id.to_string();
    
    for i in 0..count {
        let task_id = task_id.clone();
        // Rust can run all these tasks in parallel across CPU cores
        let handle = tokio::task::spawn_blocking(move || {
            // Generate large content (parallel across tasks)
            let content = generate_large_content(size_mb, &format!("{}_{}", task_id, i));
            
            // Compute hash (parallel across tasks)
            let hash = compute_hash(&content);
            
            // Compress data (parallel across tasks)
            let compressed = compress_data(&content)?;
            
            // Create file data structure
            let file_data = FileData {
                hash,
                data: hex::encode(&compressed),
                timestamp: std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_millis(),
            };
            
            // Serialize to JSON
            let json_data = serde_json::to_string(&file_data)?;
            
            // Write to file
            let filename = format!("temp_{}_{}.dat", task_id, i);
            std::fs::write(filename, json_data)?;
            
            Ok::<(), Box<dyn std::error::Error + Send + Sync>>(())
        });
        handles.push(handle);
    }
    
    for handle in handles {
        handle.await??;
    }
    
    Ok(())
}

async fn read_test_files(
    count: usize, 
    task_id: &str
) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>> {
    let mut handles: Vec<tokio::task::JoinHandle<Result<String, Box<dyn std::error::Error + Send + Sync>>>> = Vec::new();
    let task_id = task_id.to_string();
    
    for i in 0..count {
        let task_id = task_id.clone();
        // Rust runs these in parallel across CPU cores
        let handle = tokio::task::spawn_blocking(move || -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
            let filename = format!("temp_{}_{}.dat", task_id, i);
            
            // Retry logic
            for retry in 0..10 {
                match std::fs::read_to_string(&filename) {
                    Ok(json_data) => {
                        // Parse JSON
                        let file_data: FileData = serde_json::from_str(&json_data)?;
                        
                        // Decode hex data
                        let compressed = hex::decode(&file_data.data)?;
                        
                        // Decompress (parallel across tasks)
                        let content = decompress_data(&compressed)?;
                        
                        // Verify hash (parallel across tasks)
                        let hash = compute_hash(&content);
                        if hash != file_data.hash {
                            return Err("Hash mismatch".into());
                        }
                        
                        return Ok(content);
                    }
                    Err(_) if retry < 9 => {
                        std::thread::sleep(Duration::from_millis(20));
                    }
                    Err(e) => return Err(e.into()),
                }
            }
            unreachable!()
        });
        handles.push(handle);
    }
    
    let mut results = Vec::new();
    for handle in handles {
        results.push(handle.await??);
    }
    
    Ok(results)
}

async fn cleanup_test_files(count: usize, task_id: &str) {
    let mut handles = Vec::new();
    let task_id = task_id.to_string();
    
    for i in 0..count {
        let task_id = task_id.clone();
        let handle = tokio::spawn(async move {
            let filename = format!("temp_{}_{}.dat", task_id, i);
            let _ = tokio::fs::remove_file(filename).await;
        });
        handles.push(handle);
    }
    
    for handle in handles {
        let _ = handle.await;
    }
}

async fn simulate_network_delay(ms: u64) {
    sleep(Duration::from_millis(ms)).await;
}

pub struct TaskResult {
    pub files_processed: usize,
    pub task_time: Duration,
}

pub async fn io_intensive_task(task_id: &str) -> Result<TaskResult, Box<dyn std::error::Error + Send + Sync>> {
    // Reduced file count but increased file size for CPU-intensive operations
    // Rust excels with parallel CPU-bound tasks and efficient memory management
    let file_count = 10;  // Fewer files but larger and more CPU-intensive
    let file_size_mb = 2; // 2MB per file (total 20MB per task)
    let network_calls = 5;
    
    let task_start = Instant::now();
    
    // Create large files with compression and hashing
    // Rust runs these in parallel across available CPU cores
    create_test_files(file_count, task_id, file_size_mb).await?;
    
    // Short wait to ensure filesystem writes complete
    sleep(Duration::from_millis(5)).await;
    
    // Simulate network calls with delays
    let network_handles: Vec<_> = (0..network_calls)
        .map(|_| {
            let delay = 10 + rand::random::<u64>() % 20;
            tokio::spawn(simulate_network_delay(delay))
        })
        .collect();
    
    // Read, decompress, and verify files (all in parallel)
    let file_contents = read_test_files(file_count, task_id).await?;
    
    // Wait for network calls to complete
    for handle in network_handles {
        handle.await?;
    }
    
    // Cleanup
    cleanup_test_files(file_count, task_id).await;
    
    let task_end = Instant::now();
    
    Ok(TaskResult {
        files_processed: file_contents.len(),
        task_time: task_end - task_start,
    })
}

pub async fn run_io_benchmark_heavy(task_count: usize) {
    println!("Rust I/O benchmark start (Heavy CPU-intensive version)");
    println!("WHY RUST IS FASTEST:");
    println!("1. Zero-cost abstractions - no runtime overhead");
    println!("2. True parallelism with async/await + thread pool");
    println!("3. Highly optimized native code with LLVM");
    println!("4. Precise memory control with no GC pauses");
    println!("---");
    
    let start = Instant::now();
    
    let mut handles = Vec::new();
    for i in 0..task_count {
        let handle = tokio::spawn(async move {
            let task_id = format!("task{}", i);
            match io_intensive_task(&task_id).await {
                Ok(result) => {
                    println!("Task {} done in {:?}", i + 1, result.task_time);
                    Some(result)
                }
                Err(e) => {
                    println!("Task {} failed: {}", i + 1, e);
                    None
                }
            }
        });
        handles.push(handle);
    }
    
    let mut results = Vec::new();
    for handle in handles {
        if let Ok(Some(result)) = handle.await {
            results.push(result);
        }
    }
    
    let elapsed = start.elapsed();
    
    // Calculate statistics
    let total_task_time: Duration = results.iter().map(|r| r.task_time).sum();
    let avg_task_time = total_task_time / results.len() as u32;
    let total_files: usize = results.iter().map(|r| r.files_processed).sum();
    
    println!("---");
    println!("All {} tasks done in {:?}", task_count, elapsed);
    println!("Average task time: {:?}", avg_task_time);
    println!("Total files processed: {}", total_files);
}
